Using `do` in lein `:repl-options` leads to error

Based on the README (https://github.com/nrepl/piggieback#usage), it seems like I need to execute `lein repl`, then:
```
(require 'cljs.repl.node)
(cider.piggieback/cljs-repl (cljs.repl.node/repl-env))
```
in order to get a cljs repl I can interact with.

I attempted to do this by adding this to my `project.clj`:

```
  :repl-options {:init (do (require 'cljs.repl.node)
                           (cider.piggieback/cljs-repl (cljs.repl.node/repl-env)))
                 :nrepl-middleware [cider.piggieback/wrap-cljs-repl]}
```

However, when I do this, I get:

```
> lein repl
Exception in thread "main" java.lang.IllegalStateException: Can't change/establish root binding of: *cljs-repl-env* with set, compiling:(/tmp/form-init6720131333940924525.clj:1:73)
	at clojure.lang.Compiler.load(Compiler.java:7526)
	at clojure.lang.Compiler.loadFile(Compiler.java:7452)
	at clojure.main$load_script.invokeStatic(main.clj:278)
	at clojure.main$init_opt.invokeStatic(main.clj:280)
	at clojure.main$init_opt.invoke(main.clj:280)
	at clojure.main$initialize.invokeStatic(main.clj:311)
	at clojure.main$null_opt.invokeStatic(main.clj:345)
	at clojure.main$null_opt.invoke(main.clj:342)
	at clojure.main$main.invokeStatic(main.clj:424)
	at clojure.main$main.doInvoke(main.clj:387)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:702)
	at clojure.main.main(main.java:37)
Caused by: java.lang.IllegalStateException: Can't change/establish root binding of: *cljs-repl-env* with set
	at clojure.lang.Var.set(Var.java:223)
	at cider.piggieback$cljs_repl.invokeStatic(piggieback_impl.clj:184)
	at cider.piggieback$cljs_repl.doInvoke(piggieback_impl.clj:146)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at user$eval1711.invokeStatic(form-init6720131333940924525.clj:1)
	at user$eval1711.invoke(form-init6720131333940924525.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7062)
	at clojure.lang.Compiler.eval(Compiler.java:7052)
	at clojure.lang.Compiler.eval(Compiler.java:7052)
	at clojure.lang.Compiler.eval(Compiler.java:7051)
	at clojure.lang.Compiler.load(Compiler.java:7514)
	... 12 more
```

So I've been using a shell script instead to start my repl:

```
#!/bin/zsh

(echo "(require 'cljs.repl.node) (cider.piggieback/cljs-repl (cljs.repl.node/repl-env))"; cat <&0) | lein repl
```

This seems weird to me - is this intended behavior or am I missing the proper way to set this up?

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,7 @@
 
 ## unreleased
 
+* [#124](https://github.com/nrepl/piggieback/issues/124): Throw a clear error when `cljs-repl` is invoked outside of an nREPL session (e.g. from Leiningen's `:repl-options :init`) instead of a cryptic "Can't change/establish root binding".
 * [#120](https://github.com/nrepl/piggieback/issues/120): Fix printing of unknown tagged literals (a space was missing after the tag and string data lost its quotes).
 * [#62](https://github.com/nrepl/piggieback/issues/62): Keep evaluation working after a REPL setup error by always recording the compiler env.
 * [#111](https://github.com/nrepl/piggieback/issues/111): Fix ClojureScript output being tagged with the REPL-starting message's id and vanishing after reconnecting to a session.

--- a/src/cider/piggieback_impl.clj
+++ b/src/cider/piggieback_impl.clj
@@ -193,6 +193,20 @@
   "Starts a ClojureScript REPL over top an nREPL session.  Accepts
    all options usually accepted by e.g. cljs.repl/repl."
   [repl-env & {:as options}]
+  ;; cljs-repl stores its state with `set!` on dynamic vars that the nREPL
+  ;; session middleware establishes per message. Called outside of a session -
+  ;; e.g. from Leiningen's :repl-options :init, which runs at startup before any
+  ;; session exists - those vars have no thread binding and `set!` fails with a
+  ;; cryptic "Can't change/establish root binding" error (issue #124). Fail fast
+  ;; with an actionable message instead.
+  (when-not (thread-bound? #'*cljs-repl-env*)
+    (throw (ex-info
+            (str "cider.piggieback/cljs-repl must be invoked from within an nREPL "
+                 "session, e.g. by evaluating it at a REPL connected to an nREPL "
+                 "server that has the wrap-cljs-repl middleware. It cannot be "
+                 "started from a plain Clojure REPL or from Leiningen's "
+                 ":repl-options :init, both of which run outside of any session.")
+            {:repl-env repl-env})))
   (try
     (let [repl-opts (cljs.repl/-repl-options repl-env)
           repl-env (delegating-repl-env repl-env)

--- a/test/cider/piggieback_no_session_test.clj
+++ b/test/cider/piggieback_no_session_test.clj
@@ -1,0 +1,23 @@
+(ns cider.piggieback-no-session-test
+  "Regression test for https://github.com/nrepl/piggieback/issues/124: calling
+  cljs-repl outside of an nREPL session should fail with a clear message rather
+  than a cryptic \"Can't change/establish root binding\" error."
+  (:require
+   [clojure.test :refer [deftest is]]))
+
+(require 'cider.piggieback 'cljs.repl)
+
+(defn- fake-repl-env []
+  (reify cljs.repl/IJavaScriptEnv
+    (-setup [_ _] nil)
+    (-evaluate [_ _ _ _] {:status :success :value "nil"})
+    (-load [_ _ _] nil)
+    (-tear-down [_] nil)))
+
+(deftest cljs-repl-outside-session-fails-clearly
+  ;; This test thread is not an nREPL session, so the piggieback vars are not
+  ;; thread-bound - the same situation as a Leiningen :init.
+  (is (thrown-with-msg?
+       clojure.lang.ExceptionInfo
+       #"must be invoked from within an nREPL session"
+       (cider.piggieback/cljs-repl (fake-repl-env)))))

